### PR TITLE
Rerun failed tests automatically

### PR DESCRIPTION
**Pull request by bergundy**
## What changed?

Leverage gotestsum to automatically retry failed tests in all suites (unit, integration, functional, xdc, ndc).
By default, tests will be retried twice if there are no more than 10 failed tests in the suite.

### Changes to the `Makefile`

Add `FAILED_TEST_RETRIES` variable, defaulting to `2`.

Removed the use of `SINGLE_TEST_ARGS`, `TEST_ARGS` can be used instead.
Moved `-race` flag to its own separate flag and freed up `TEST_ARGS` to be set by callers without.

Added:
```
# Whether or not to test with the race detector. All of (1 on y yes t true) are true values.
TEST_RACE_FLAG ?= on
# Whether or not to shuffle tests. All of (1 on y yes t true) are true values.
TEST_SHUFFLE_FLAG ?= on
```

Added `COMPILED_TEST_ARGS` that combines the common flags and options to be reused by all test invocations.

### Changes to `run-tests.yml`

Removed the use of `SINGLE_TEST_ARGS`.

Merged the `Run single unit test` step into `Run unit tests`.

Set `FAILED_TEST_RETRIES: "0"` on flaky test detection steps.

Removed steps to upload test results to buildkite and relevant script / Makefile content.

Moved flaky test detection to after the upload test results (to GH) step in all jobs.

## Why?

So we can enable flaky tests to get coverage without hurting the CI experience.

### CODE REVIEW

File: .github/workflows/run-tests.yml [hunk 1]
```diff
@@ -260,21 +259,21 @@ jobs:
...
+        env:
+          UNIT_TEST_DIRS: ${{ inputs.unit_test_directory }}
+          TEST_ARGS: ${{ needs.set-up-single-test.outputs.single_test_args }}
 
       - name: Flaky Unit Test Detection
         if: ${{ !cancelled() && !inputs.run_single_unit_test && env.MODIFIED_TEST_SUITES != '' }}
         timeout-minutes: 30
         run: |
           echo "Detecting flaky unit tests: ${{ needs.set-up-single-test.outputs.modified_unit_test_suites }}"
           make unit-test TEST_ARGS="-run=${{ needs.set-up-single-test.outputs.modified_unit_test_suites }} -count=10"
```
Comment (dnr) [inline .github/workflows/run-tests.yml:273]: put this TEST_ARGS in the env: also? (for consistency and ease of reading)

File: .github/workflows/run-tests.yml [hunk 2]
```diff
@@ -390,12 +393,11 @@ jobs:
       TEST_SHARD_INDEX: ${{ matrix.shard_index }}
       PERSISTENCE_TYPE: ${{ matrix.persistence_type }}
       PERSISTENCE_DRIVER: ${{ matrix.persistence_driver }}
-      SINGLE_TEST_ARGS: ${{ needs.set-up-single-test.outputs.single_test_args }}
       TEST_TIMEOUT: ${{ needs.set-up-single-test.outputs.test_timeout }}
       BUILDKITE_MESSAGE: '{"job": "functional-test", "db": "${{ matrix.persistence_driver }}"}'
     steps:
       - name: Print functional test
-        run: echo "${{ needs.set-up-single-test.outputs.dbs }}" && echo "$SINGLE_TEST_ARGS"
+        run: echo "${{ needs.set-up-single-test.outputs.dbs }}" && echo "${{ needs.set-up-single-test.outputs.single_test_args }}"
```
Comment (dnr) [inline .github/workflows/run-tests.yml:400]: do we really need this? why not just have the thing that runs functional tests log the exact command it's using

File: .github/workflows/run-tests.yml [hunk 3]
```diff
@@ -436,14 +438,17 @@ jobs:
       - name: Run functional test
         if: ${{ inputs.run_single_functional_test != true || (inputs.run_single_functional_test == true && contains(fromJSON(needs.set-up-single-test.outputs.dbs), env.PERSISTENCE_DRIVER)) }}
         timeout-minutes: ${{ fromJSON(needs.set-up-single-test.outputs.github_timeout) }} # make sure this is larger than the test timeout in the Makefile
-        run: make functional-test-coverage
+        run: make functional-test-coverage TEST_ARGS="${{ needs.set-up-single-test.outputs.single_test_args }}"
```
Comment (dnr) [inline .github/workflows/run-tests.yml:441]: use env: here too? (and the next three below)

File: Makefile [hunk 4]
```diff
@@ -46,13 +45,35 @@ ALL_TEST_TAGS := $(ALL_BUILD_TAGS),$(TEST_TAG)
...
+# NOTE: This is incompatible with TEST_ARGS which specify the `-run` flag due to how gotestsum selects which tests to
+# retry.
+FAILED_TEST_RETRIES ?= 2
+
+# Whether or not to test with the race detector. All of (1 on y yes t true) are true values.
+TEST_RACE_FLAG ?= on
+# Whether or not to shuffle tests. All of (1 on y yes t true) are true values.
+TEST_SHUFFLE_FLAG ?= on
+# Common test args used in the various test suite targets.
+COMPILED_TEST_ARGS := -timeout=$(TEST_TIMEOUT) \
```
Comment (dnr) [inline Makefile:64]: Hmm.. should we use TEST_TIMEOUT here? This is used for unit tests too and they have a different timeout in the GHA workflow...

File: Makefile [hunk 5]
```diff
@@ -382,36 +399,49 @@ $(TEST_OUTPUT_ROOT):
 
 prepare-coverage-test: $(GOTESTSUM) $(TEST_OUTPUT_ROOT)
 
+# Runs gotestsum, retrying failed tests as specified by the FAILED_TEST_RETRIES variable.
+# A function is required here there are two variants of this command, putting the packages to test in different
+# positions.
+# NOTE: Packages need to be provided via a named argument when --rerun-fails is on as stated in the gotestsum README.
+# NOTE: This had to be a long line to avoid adding newlines to the generated command.
+define gotestsum_coverage
+	$(if $(shell test $(FAILED_TEST_RETRIES) -eq 0 && echo "OK"), $(GOTESTSUM) --junitfile $(NEW_REPORT) -- $(1), $(GOTESTSUM) --rerun-fails=$(FAILED_TEST_RETRIES) --rerun-fails-max-failures=10 --junitfile $(NEW_REPORT) --packages $(1) --)
```
Comment (dnr) [inline Makefile:408]: first, can we split this onto a few lines for readability? I see the comment but there's got to be a way to do it with backslashes or something.

second, using shell to check a make/env variable seems unnecessary. why not use make's `ifeq` (with two versions of the function in the two branches)?

third, why not just use `--packages` all the time? isn't that simpler?

File: Makefile [hunk 6]
```diff
@@ -382,36 +399,49 @@ $(TEST_OUTPUT_ROOT):
...
 pre-build-functional-test-coverage: prepare-coverage-test
 	go test -c -o /dev/null $(FUNCTIONAL_TEST_ROOT) $(TEST_ARGS) $(TEST_TAG_FLAG) $(FUNCTIONAL_TEST_COVERPKG)
 
 functional-test-coverage: prepare-coverage-test
 	@printf $(COLOR) "Run functional tests with coverage with $(PERSISTENCE_DRIVER) driver..."
-	$(GOTESTSUM) --junitfile $(NEW_REPORT) -- \
-		$(FUNCTIONAL_TEST_ROOT) -shuffle on -timeout=$(TEST_TIMEOUT) $(TEST_ARGS) $(SINGLE_TEST_ARGS) $(TEST_TAG_FLAG) \
-		 -persistenceType=$(PERSISTENCE_TYPE) -persistenceDriver=$(PERSISTENCE_DRIVER) \
-		 $(FUNCTIONAL_TEST_COVERPKG) -coverprofile=$(NEW_COVER_PROFILE) $(TEST_PARALLEL_FLAGS)
+	cd $(FUNCTIONAL_TEST_ROOT) && \
```
Comment (dnr) [inline Makefile:427]: why cd? that shouldn't be necessary

File: Makefile (hunk 5 above)
Comment (bergundy) [reply Makefile]: I thought `--packages` was incompatible with `TEST_ARGS="-run ..."` but I was wrong. Simplified.

File: Makefile [hunk 7]
```diff
@@ -352,26 +369,26 @@ build-tests:
 
 unit-test: clean-test-results
 	@printf $(COLOR) "Run unit tests..."
-	@go test $(UNIT_TEST_DIRS) -shuffle on -timeout=$(TEST_TIMEOUT) $(TEST_TAG_FLAG) $(TEST_ARGS) 2>&1 | tee -a test.log
+	@go test $(UNIT_TEST_DIRS) $(COMPILED_TEST_ARGS) 2>&1 | tee -a test.log
```
Comment (dnr) [inline Makefile:372]: personally I'd rather see the full command here and everywhere, so it's easy to copy-paste with an extra `-run` or whatever to narrow down to one test, but be confident I'm using the right flags

File: Makefile [hunk 8]
```diff
@@ -384,34 +401,40 @@ prepare-coverage-test: $(GOTESTSUM) $(TEST_OUTPUT_ROOT)
...
 functional-test-coverage: prepare-coverage-test
 	@printf $(COLOR) "Run functional tests with coverage with $(PERSISTENCE_DRIVER) driver..."
-	$(GOTESTSUM) --junitfile $(NEW_REPORT) -- \
-		$(FUNCTIONAL_TEST_ROOT) -shuffle on -timeout=$(TEST_TIMEOUT) $(TEST_ARGS) $(SINGLE_TEST_ARGS) $(TEST_TAG_FLAG) \
-		 -persistenceType=$(PERSISTENCE_TYPE) -persistenceDriver=$(PERSISTENCE_DRIVER) \
-		 $(FUNCTIONAL_TEST_COVERPKG) -coverprofile=$(NEW_COVER_PROFILE) $(TEST_PARALLEL_FLAGS)
+	$(GOTESTSUM) --rerun-fails=$(FAILED_TEST_RETRIES) --rerun-fails-max-failures=10 --junitfile $(NEW_REPORT) --packages $(FUNCTIONAL_TEST_ROOT) -- \
+		$(COMPILED_TEST_ARGS) \
+		-coverprofile=$(NEW_COVER_PROFILE) $(FUNCTIONAL_TEST_COVERPKG) \
+		-args -persistenceType=$(PERSISTENCE_TYPE) -persistenceDriver=$(PERSISTENCE_DRIVER)
```
Comment (dnr) [inline Makefile:423]: ah nice, I didn't know `-args`. I like the clean separation

**Review by dnr — APPROVED** [review]

File: .github/workflows/run-tests.yml [hunk 9]
```diff
@@ -260,35 +256,35 @@ jobs:
           key: go-${{ runner.os }}${{ runner.arch }}-build-${{ env.COMMIT }}
 
       - name: Run unit tests
-        if: ${{ !cancelled() && !inputs.run_single_unit_test }}
         timeout-minutes: 15
```
[the remark this answers is not in the document]
Comment (bergundy) [reply .github/workflows/run-tests.yml:259]: This PR already does a lot more than I initially intended. Let's revisit this.

I'll open an issue.

File: .github/workflows/run-tests.yml [hunk 10]
```diff
@@ -260,35 +256,35 @@ jobs:
...
+          UNIT_TEST_DIRS: ${{ inputs.unit_test_directory }}
+          TEST_ARGS: ${{ needs.set-up-single-test.outputs.single_test_args }}
+          TEST_TIMEOUT: ${{ needs.set-up-single-test.outputs.test_timeout }}
 
       - name: Upload test results
-        if: ${{ !cancelled() }}
-        run: make upload-test-results
-
-      - name: Upload test results as a GitHub Actions artifact
         uses: actions/upload-artifact@v4.4.0
```
[the remark this answers is not in the document]
Comment (bergundy) [reply .github/workflows/run-tests.yml:267]: I'll do this in the next PR.

File: Makefile (hunk 4 above)
Comment (bergundy) [reply Makefile:64]: Ah, okay, I thought you were referring to a change I made.

File: Makefile (hunk 7 above)
Comment (bergundy) [reply Makefile:372]: Same, I'll raise this with the team and confirm there's no objections to change.

File: Makefile (hunk 8 above)
Comment (bergundy) [reply Makefile:423]: Me neither but apparently we had to use this for gotestsum to properly construct the command.

**Pull request merged**